File: services/typescript/fastify/src/routes/handlers/post.ts
import type { RouteHandler } from "fastify";
import { randomUUID } from "node:crypto";
import { type State, parseTask } from "app";
import { getLogger } from "telemetry";
import { AppErrors, sendError } from "../errors.js";
import type { PostRoute } from "../types.js";

export function postHandler(state: State): RouteHandler<PostRoute> {
  return async (request, reply) => {
    const logger = getLogger();

    let task = parseTask(request.body);
    if (!task.success) {
      logger.warn(
        { method: request.method, path: request.url },
        "Invalid body JSON"
      );
      return sendError(reply, AppErrors.InvalidJsonBody);
    }

    let id = randomUUID();

    state.tasks.set(id, task.data);

    logger.info(
      {
        id,
        secret: task.data.secret.length,
        operation: task.data.operation.toString().toLowerCase(),
        method: "POST"
      },
      "Inserted new task"
    );

    return reply.status(201).send({ id });
  };
}

File: services/python/quart/src/routes/handlers/post.py
import structlog
from uuid import uuid4, UUID
from quart import current_app, jsonify, request
from pydantic import ValidationError

from app.task import Task
from app.params import parse_task
from app.state import AppState
from routes.errors import send_error, AppErrors


async def post_handler():
    logger = structlog.get_logger()

    body = await request.get_json(silent=True)

    if body is None:
        logger.warn("Invalid body JSON", method=request.method, path=request.path)
        return send_error(AppErrors.InvalidJsonBody)

    try:
        task: Task = parse_task(body)
    except ValidationError:
        logger.warn("Invalid body JSON", method=request.method, path=request.path)
        return send_error(AppErrors.InvalidJsonBody)

    id: UUID = uuid4()
    state: AppState = current_app.extensions["state"]

    state.tasks[id] = task

    logger.info(
        "Inserted new task",
        id=str(id),
        secret=len(task.secret),
        operation=task.operation.lower(),
        method=request.method,
    )

    return jsonify({"id": str(id)}), 201
